src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,94 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
-    @Override
-    public void add(T value) {
+    private static final int INITIALIZATION_SIZE = 10;
+    private static final double ARRAY_RATIO = 1.5;
+    private T[] values;
+    private int size;
 
+    public ArrayList() {
+        values = (T[]) new Object[INITIALIZATION_SIZE];
     }
 
     @Override
-    public void add(T value, int index) {
+    public void add(T value) {
+        growIfArrayFull();
+        values[size] = value;
+        size++;
+    }
 
+    public void add(T value, int index) {
+        checkIndex(index, size);
+        T[] newValues = Arrays.copyOf(values, values.length + 1);
+        System.arraycopy(newValues, index, newValues, index + 1, size - index);
+        newValues[index] = value;
+        values = newValues;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        growIfArrayFull();
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, size - 1);
+        return values[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index,size - 1);
+        values[index] = value;
     }
 
-    @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, size - 1);
+        T[] newValues = Arrays.copyOf(values, values.length - 1);
+        System.arraycopy(values, index + 1, newValues, index, values.length - index - 1);
```

Comment:
```suggestion
        System.arraycopy(values, index + 1, values, index, values.length - index - 1);
```
you can use it twice